Query related records

mpayson opened the pull request:
Hey arcgis-rest-js team!

`queryRelatedRecords` came up in a project so I thought I'd try TS and write a wrapper--please let me know if you have feedback/suggested changes

Also, how do I compile and test the function? I wrote a couple mock [unit tests](https://github.com/mpayson/arcgis-rest-js/blob/mp-queryrelates/packages/arcgis-rest-feature-service/test/features.test.ts#L175) following the `queryFeatures` [pattern](https://github.com/mpayson/arcgis-rest-js/blob/mp-queryrelates/packages/arcgis-rest-feature-service/test/features.test.ts#L45), but I'm not sure how to access this function to test on a live feature service

thanks!

cc @jgravois

jgravois commented on packages/arcgis-rest-feature-service/src/queryRelated.ts:
```diff
@@ -0,0 +1,99 @@
+
+  appendCustomParams(requestOptions, options);
+
+  if (!options.params.definitionExpression) {
```
while we're stubbing in 'sensible' defaults...
```js
if (!options.params.relationshipId) {
  options.params.relationshipId = 0; // this is falsy, but the internal method will catch that.
}
```

jgravois commented on packages/arcgis-rest-feature-service/src/queryRelated.ts:
```diff
@@ -0,0 +1,99 @@
+/* Copyright (c) 2017 Environmental Systems Research Institute, Inc.
```
2018

jgravois commented on packages/arcgis-rest-feature-service/src/queryRelated.ts:
```diff
@@ -0,0 +1,99 @@
+  returnTrueCurves?: boolean;
+  orderByFields?: string;
+  returnCountOnly?: boolean;
+}
```
i'm starting to think that we are better served to _only promote the absolutely most _common_ parameters as "first class citizens" and instead just show code snippets that prove there is an underlying mechanism to pass in the more obscure params.

```ts
export interface IQueryRelatedRecordsRequestOptions extends IRequestOptions {
  url: string;
  relationshipId?: number;
  objectIds?: number[];
  outFields?: "*" | string[];
}

/**
 * Query the related records for a feature service. See [REST Documentation](https://developers.arcgis.com/rest/services-reference/query-related-records-feature-service-.htm) for more information.
 *
 * ```js
 * import { queryRelatedRecords } from '@esri/arcgis-rest-feature-service'
 *
 * const url = "http://services.myserver/OrgID/ArcGIS/rest/services/Petroleum/KSPetro/FeatureServer/0"
 *
 * queryRelatedRecords({
 *  url: url,
 *  relationshipId: 1,
 *  params: { returnGeometry: false }
 * })
 *  .then(response => {
 *    console.log(response.relatedRecords)
 *  })

 * ```

```
doing it this way is less work and i think it makes our own documentation less intimidating.

mpayson replied:
I like that approach. Are`url` `relationshipId` `objectIds` `outFields` the most common? What about `definitionExpression` `returnGeometry` `returnCountOnly`? I mention those only because I've actually used them

mpayson replied:
also updating the doc to read:
> Query related records request options. Additional options can be passed as [params](https://esri.github.io/arcgis-rest-js/api/request/IRequestOptions/#params) from [IRequestOptions](https://esri.github.io/arcgis-rest-js/api/request/IRequestOptions/). See the [REST Documentation](https://developers.arcgis.com/rest/services-reference/query-related-records-feature-service-.htm) for more information and a full list of parameters.

jgravois replied:
i'm going to just embrace the arbitrary nature of my code review and say this: if you've actually used it, thats common enough for me.

lets include all the parameters you listed in the interface _except_ for `returnGeometry` (and use that one in the code snippet to demo the method's extensibility).

I think that your idea of being more explicit in the doc about this is a really good idea too. how about this (emphasis mine):

> Related record query request options. Additional **arguments** can be passed **via the params property**. See the REST Documentation for more information and a full list of parameters.

jgravois commented on packages/arcgis-rest-feature-service/src/queryRelated.ts:
```diff
@@ -0,0 +1,99 @@
+ * queryRelatedRecords({
+ *  url: url,
+ *  relationshipId: 1
+ * };)
```
~~`};)`~~ `})`

mpayson commented:
* [x] tested with this sample [data](https://services.arcgis.com/q7zPNeKmTWeh7Aor/arcgis/rest/services/service_c3d3cd8fb4ff47d1be602597bb778d39/FeatureServer/0) from "Office Hours" app and it works
* [x] `};)` -> `})`
* [x] ~2017~ -> 2018
* [x] default `relationshipId` + updated test
* [x] promote common parameters
  * [x] removed `returnCountOnly` and put it in the example (to reflect `count` in `IRelatedRecordGroup`)
  * [x] updated doc

jgravois commented:
released in [`v1.7.0`](https://github.com/Esri/arcgis-rest-js/releases/tag/v1.7.0) 🎉
